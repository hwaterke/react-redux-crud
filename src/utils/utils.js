// @flow

export function invariant(condition: any, message: string) {
  if (!condition) {
    const error = new Error(message)
    error.name = 'Invariant Violation'
    throw error
  }
}
