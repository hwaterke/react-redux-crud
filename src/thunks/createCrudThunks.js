// @flow
import reduxCrud from 'redux-crud'
import axios from 'axios'
import type {ResourceDefinition} from '../types/ResourceDefinition'
import {invariant, requiredParam} from '../utils/utils'

function dataOrError(error) {
  if (error.response && error.response.data) {
    return error.response.data
  }
  return error
}

export type CrudConfig = {
  /**
   * An optional function that returns the base url of the backend.
   * First argument is the redux state.
   * Default is an empty string
   */
  backendSelector?: (state: any) => string,

  /**
   * An optional function that returns headers to use in every request.
   * First argument is the redux state.
   * Default is an empty object
   * This can be used to add an Authorization token which is stored in redux.
   */
  headersSelector?: (state: any) => {[headerName: string]: string},

  /**
   * This optional function is called whenever an error occurs.
   */
  onError?: (
    resource: ResourceDefinition,
    operation: string,
    error: any,
    dispatch: Function
  ) => void,

  /**
   * This mandatory function is used to generated a unique id for a entity being created.
   * This is used for optimistic creation
   */
  cuid: () => string,

  /**
   * This optional function is used to select the record array from a fetchAll response from the backend.
   * The first parameter is the body of the response.
   * The default implementation is the identity function
   */
  fetchAllDataToRecords?: (responseData: any) => any[],

  /**
   * This optional function is used to select the record from a fetchOne response from the backend.
   * The first parameter is the body of the response.
   * The default implementation is the identity function
   */
  fetchOneDataToRecord?: (responseData: any) => any,

  /**
   * This optional function is used to select the record from a create response from the backend.
   * The first parameter is the body of the response.
   * The default implementation is the identity function
   */
  createDataToRecord?: (responseData: any) => any,

  /**
   * This optional function is used to select the record from an update response from the backend.
   * The first parameter is the body of the response.
   * The default implementation is the identity function
   */
  updateDataToRecord?: (responseData: any) => any,
}

const defaultConfiguration = {
  backendSelector: (state: any) => '',
  headersSelector: (state: any) => ({}),
  onError: (
    resource: ResourceDefinition,
    operation: string,
    error: any,
    dispatch: Function
  ) => {},
  fetchAllDataToRecords: data => data,
  fetchOneDataToRecord: data => data,
  createDataToRecord: data => data,
  updateDataToRecord: data => data,
}

export function createCrudThunks(configuration: CrudConfig) {
  const config = {
    ...defaultConfiguration,
    ...configuration,
  }

  invariant(
    config.cuid,
    'You need to specify a cuid function in your CRUD configuration'
  )

  return {
    fetchAll: ({
      resource = requiredParam('resource'),
      path,
      replace = false,
      params,
    }: {
      resource: ResourceDefinition,
      path?: string,
      replace?: boolean,
      params?: {[string]: string},
    }) => (dispatch: Function, getState: Function) => {
      const resourcePath = path || resource.defaultPath

      invariant(resourcePath, 'You need to specify a path to fetch records')

      const baseActionCreators = reduxCrud.actionCreatorsFor(resource.name, {
        key: resource.key,
      })

      dispatch(baseActionCreators.fetchStart())

      const promise = axios({
        url: `${config.backendSelector(getState())}/${resourcePath}`,
        method: 'get',
        headers: config.headersSelector(getState()),
        params,
      })

      promise
        .then(
          response => {
            dispatch(
              baseActionCreators.fetchSuccess(
                config.fetchAllDataToRecords(response.data),
                {replace}
              )
            )
          },
          error => {
            dispatch(baseActionCreators.fetchError(dataOrError(error)))
            config.onError(resource, 'fetchAll', error, dispatch)
          }
        )
        .catch(err => {
          config.onError(resource, 'fetchAll', err, dispatch)
        })

      return promise
    },

    fetchOne: ({
      resource = requiredParam('resource'),
      path,
      uuid = requiredParam('uuid'),
      params,
    }: {
      resource: ResourceDefinition,
      path?: string,
      uuid: string,
      params?: {[string]: string},
    }) => (dispatch: Function, getState: Function) => {
      const resourcePath =
        path ||
        (resource.defaultPath ? `${resource.defaultPath}/${uuid}` : null)
      invariant(resourcePath, 'You need to specify a path to fetch a record')

      const baseActionCreators = reduxCrud.actionCreatorsFor(resource.name, {
        key: resource.key,
      })

      dispatch(baseActionCreators.fetchStart({fetchOne: uuid}))

      const promise = axios({
        url: `${config.backendSelector(getState())}/${resourcePath}`,
        method: 'get',
        headers: config.headersSelector(getState()),
        params,
      })

      promise
        .then(
          response => {
            dispatch(
              baseActionCreators.fetchSuccess(
                config.fetchOneDataToRecord(response.data),
                {fetchOne: uuid}
              )
            )
          },
          error => {
            dispatch(
              baseActionCreators.fetchError(dataOrError(error), {
                fetchOne: uuid,
              })
            )
            config.onError(resource, 'fetchOne', error, dispatch)
          }
        )
        .catch(err => {
          config.onError(resource, 'fetchOne', err, dispatch)
        })

      return promise
    },

    createResource: ({
      resource = requiredParam('resource'),
      path,
      entity = requiredParam('entity'),
      optimistic = true,
    }: {
      resource: ResourceDefinition,
      path?: string,
      entity: any,
      optimistic: boolean,
    }) => (dispatch: Function, getState: Function) => {
      const resourcePath = path || resource.defaultPath
      invariant(resourcePath, 'You need to specify a path to create a record')

      // Create a client id
      const cid = entity[resource.key] || config.cuid()
      if (!entity[resource.key]) {
        entity = {...entity, [resource.key]: cid}
      }

      const baseActionCreators = reduxCrud.actionCreatorsFor(resource.name, {
        key: resource.key,
      })

      if (optimistic) {
        dispatch(baseActionCreators.createStart(entity))
      }

      const promise = axios({
        url: `${config.backendSelector(getState())}/${resourcePath}`,
        method: 'post',
        headers: config.headersSelector(getState()),
        data: entity,
      })

      promise
        .then(
          response => {
            dispatch(
              baseActionCreators.createSuccess(
                config.createDataToRecord(response.data),
                cid
              )
            )
          },
          error => {
            dispatch(baseActionCreators.createError(dataOrError(error), entity))
            config.onError(resource, 'create', error, dispatch)
          }
        )
        .catch(err => {
          config.onError(resource, 'create', err, dispatch)
        })

      return promise
    },

    updateResource: ({
      resource = requiredParam('resource'),
      path,
      merge = true,
      entity = requiredParam('entity'),
      method = 'patch',
      body = null,
      optimistic = true,
    }: {
      resource: ResourceDefinition,
      path?: string,
      merge?: boolean,
      entity: any,
      method: string,
      body: any,
      optimistic: boolean,
    }) => (dispatch: Function, getState: Function) => {
      const resourcePath =
        path ||
        (resource.defaultPath
          ? `${resource.defaultPath}/${entity[resource.key]}`
          : null)
      invariant(resourcePath, 'You need to specify a path to update a record')
      invariant(
        entity[resource.key],
        'The entity needs to have a key to update a record'
      )

      const baseActionCreators = reduxCrud.actionCreatorsFor(resource.name, {
        key: resource.key,
      })

      if (optimistic) {
        dispatch(baseActionCreators.updateStart(entity, {merge}))
      }

      const promise = axios({
        url: `${config.backendSelector(getState())}/${resourcePath}`,
        method,
        headers: config.headersSelector(getState()),
        data: body || entity,
      })

      promise
        .then(
          response => {
            dispatch(
              baseActionCreators.updateSuccess(
                config.updateDataToRecord(response.data)
              )
            )
          },
          error => {
            dispatch(baseActionCreators.updateError(dataOrError(error), entity))
            config.onError(resource, 'update', error, dispatch)
          }
        )
        .catch(err => {
          config.onError(resource, 'update', err, dispatch)
        })

      return promise
    },

    deleteResource: ({
      resource = requiredParam('resource'),
      path,
      entity = requiredParam('entity'),
    }: {
      resource: ResourceDefinition,
      path?: string,
      entity: any,
    }) => (dispatch: Function, getState: Function) => {
      const resourcePath =
        path ||
        (resource.defaultPath
          ? `${resource.defaultPath}/${entity[resource.key]}`
          : null)
      invariant(resourcePath, 'You need to specify a path to delete a record')
      invariant(
        entity[resource.key],
        'The entity needs to have a key to delete a record'
      )

      const baseActionCreators = reduxCrud.actionCreatorsFor(resource.name, {
        key: resource.key,
      })

      dispatch(baseActionCreators.deleteStart(entity))

      const promise = axios({
        url: `${config.backendSelector(getState())}/${resourcePath}`,
        method: 'delete',
        headers: config.headersSelector(getState()),
      })

      promise
        .then(
          () => {
            dispatch(baseActionCreators.deleteSuccess(entity))
          },
          error => {
            dispatch(baseActionCreators.deleteError(dataOrError(error), entity))
            config.onError(resource, 'delete', error, dispatch)
          }
        )
        .catch(err => {
          config.onError(resource, 'delete', err, dispatch)
        })

      return promise
    },
  }
}
