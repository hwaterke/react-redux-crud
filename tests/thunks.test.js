import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import type {CrudConfig} from '../src/thunks/createCrudThunks'
import {createCrudThunks} from '../src/thunks/createCrudThunks'
import {Book, initialiseStore} from './utils'

describe('Thunks', () => {
  let store
  let intermediaryStates

  let crudConfig: CrudConfig
  let thunks
  const axiosMock = new MockAdapter(axios)

  beforeEach(() => {
    intermediaryStates = []

    store = initialiseStore()
    store.subscribe(() => intermediaryStates.push(store.getState()))

    let nextCid = 0
    crudConfig = {
      cuid: () => `cid${nextCid++}`,
    }

    thunks = createCrudThunks(crudConfig)

    axiosMock.reset()
  })

  describe('fetchAll', () => {
    it('should fetch records', async () => {
      axiosMock.onGet('/books').reply(200, [{uuid: 1, name: 'A book'}])

      await store.dispatch(thunks.fetchAll({resource: Book}))

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should merge fetched records', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 2, name: 'Another book'},
        ])
      axiosMock
        .onGet('/books2')
        .reply(200, [
          {uuid: 2, name: 'A second book'},
          {uuid: 3, name: 'A third book'},
        ])

      await store.dispatch(thunks.fetchAll({resource: Book}))
      await store.dispatch(thunks.fetchAll({resource: Book, path: 'books2'}))

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should replace fetched records', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 2, name: 'Another book'},
        ])
      axiosMock
        .onGet('/books2')
        .reply(200, [
          {uuid: 2, name: 'A second book'},
          {uuid: 3, name: 'A third book'},
        ])

      await store.dispatch(thunks.fetchAll({resource: Book}))
      await store.dispatch(
        thunks.fetchAll({resource: Book, path: 'books2', replace: true})
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should add custom headers', async () => {
      axiosMock.onGet('/books').reply(config => {
        return [200, [{uuid: 1, token: config.headers.Authorization}]]
      })

      const headerThunks = createCrudThunks({
        ...crudConfig,
        headersSelector: () => ({Authorization: '123'}),
      })

      await store.dispatch(headerThunks.fetchAll({resource: Book}))

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should handle errors', async () => {
      axiosMock.onGet('/books').networkError()

      expect.assertions(4)
      try {
        await store.dispatch(thunks.fetchAll({resource: Book}))
      } catch (e) {
        expect(e.message).toBe('Network Error')
      }

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })
  })

  describe('fetchOne', () => {
    it('should fetch one record', async () => {
      axiosMock.onGet('/books/123').reply(200, {uuid: 123, name: 'A book'})

      await store.dispatch(thunks.fetchOne({resource: Book, uuid: '123'}))

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should merge one record', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 2, name: 'Another book'},
        ])
      axiosMock.onGet('/books/2').reply(200, {uuid: 2, name: 'A second book'})

      await store.dispatch(thunks.fetchAll({resource: Book}))
      await store.dispatch(thunks.fetchOne({resource: Book, uuid: '2'}))

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should handle errors', async () => {
      axiosMock.onGet('/books/2').networkError()

      expect.assertions(4)
      try {
        await store.dispatch(thunks.fetchOne({resource: Book, uuid: '2'}))
      } catch (e) {
        expect(e.message).toBe('Network Error')
      }

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })
  })

  describe('createResource', () => {
    it('should create a record', async () => {
      axiosMock.onPost('/books').reply(201, {uuid: 1, name: 'A book'})

      await store.dispatch(
        thunks.createResource({resource: Book, entity: {name: 'A book'}})
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should handle errors', async () => {
      axiosMock.onPost('/books').networkError()

      expect.assertions(4)
      try {
        await store.dispatch(
          thunks.createResource({resource: Book, entity: {name: 'A book'}})
        )
      } catch (e) {
        expect(e.message).toBe('Network Error')
      }

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })
  })

  describe('updateResource', () => {
    it('should update a record', async () => {
      axiosMock.onPatch('/books/123').reply(200, {uuid: 123, name: 'A book'})

      await store.dispatch(
        thunks.updateResource({
          resource: Book,
          entity: {uuid: 123, name: 'A book'},
        })
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should update an existing record', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 123, name: 'Another book'},
        ])
      axiosMock
        .onPatch('/books/123')
        .reply(200, {uuid: 123, name: 'An updated book'})

      await store.dispatch(thunks.fetchAll({resource: Book}))
      await store.dispatch(
        thunks.updateResource({
          resource: Book,
          entity: {
            uuid: 123,
            name: 'An updated book',
          },
        })
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should handle errors', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 123, name: 'Another book'},
        ])
      axiosMock.onPatch('/books/123').networkError()

      expect.assertions(4)
      try {
        await store.dispatch(
          thunks.updateResource({
            resource: Book,
            entity: {
              uuid: 123,
              name: 'An updated book',
            },
          })
        )
      } catch (e) {
        expect(e.message).toBe('Network Error')
      }

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })
  })

  describe('deleteResource', () => {
    it('should delete a record', async () => {
      axiosMock.onDelete('/books/123').reply(204)

      await store.dispatch(
        thunks.deleteResource({resource: Book, entity: {uuid: 123}})
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should delete an existing record', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 123, name: 'Another book'},
        ])
      axiosMock.onDelete('/books/123').reply(204)

      await store.dispatch(thunks.fetchAll({resource: Book}))
      await store.dispatch(
        thunks.deleteResource({resource: Book, entity: {uuid: 123}})
      )

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })

    it('should handle errors', async () => {
      axiosMock
        .onGet('/books')
        .reply(200, [
          {uuid: 1, name: 'A book'},
          {uuid: 123, name: 'Another book'},
        ])
      axiosMock.onDelete('/books/123').networkError()

      expect.assertions(4)
      await store.dispatch(thunks.fetchAll({resource: Book}))
      try {
        await store.dispatch(
          thunks.deleteResource({resource: Book, entity: {uuid: 123}})
        )
      } catch (e) {
        expect(e.message).toBe('Network Error')
      }

      expect(intermediaryStates.map(s => s.lastAction)).toMatchSnapshot()
      expect(
        intermediaryStates.map(s => s.resourcesActivity.books)
      ).toMatchSnapshot()
      expect(intermediaryStates.map(s => s.resources.books)).toMatchSnapshot()
    })
  })
})
