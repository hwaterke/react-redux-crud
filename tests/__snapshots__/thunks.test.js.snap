// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thunks createResource should create a record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks deleteResource should delete a record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {},
    "categories": Object {},
  },
}
`;

exports[`Thunks deleteResource should delete an existing record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchAll should add custom headers 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "token": "123",
        "uuid": 1,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchAll should fetch records 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchAll should merge fetched records 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
      "2": Object {
        "name": "A second book",
        "uuid": 2,
      },
      "3": Object {
        "name": "A third book",
        "uuid": 3,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchAll should replace fetched records 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "2": Object {
        "name": "A second book",
        "uuid": 2,
      },
      "3": Object {
        "name": "A third book",
        "uuid": 3,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchOne should fetch one record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "123": Object {
        "name": "A book",
        "uuid": 123,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks fetchOne should merge one record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
      "2": Object {
        "name": "A second book",
        "uuid": 2,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks updateResource should update a record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "123": Object {
        "name": "A book",
        "uuid": 123,
      },
    },
    "categories": Object {},
  },
}
`;

exports[`Thunks updateResource should update an existing record 1`] = `
Object {
  "activity": Object {
    "books": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
    "categories": Object {
      "activity": 0,
      "create": 0,
      "fetch": 0,
      "remove": 0,
      "update": 0,
    },
  },
  "resources": Object {
    "books": Object {
      "1": Object {
        "name": "A book",
        "uuid": 1,
      },
      "123": Object {
        "name": "An updated book",
        "uuid": 123,
      },
    },
    "categories": Object {},
  },
}
`;
